rust/common-ifc/src/graph/port_graph.rs

```diff
@@ -0,0 +1,119 @@
+    /// A unique to the graph identifier for this node.
+    fn id(&self) -> &NodeId;
+    /// All input ports for this node.
+    fn inputs<'a>(&'a self) -> impl Iterator<Item = &'a PortName>
+    where
+        PortName: 'a;
+    /// All output ports for this node.
+    fn outputs<'a>(&'a self) -> impl Iterator<Item = &'a PortName>
+    where
+        PortName: 'a;
+}
+
+/// An edge in a [PortGraph].
+pub trait PortGraphEdge<NodeId: PortGraphId, PortName: PortGraphId>: Debug {
+    /// The source connection output port.
+    fn source(&self) -> (&NodeId, &PortName);
+    /// The target connection input port.
+    fn target(&self) -> (&NodeId, &PortName);
+}
+
+impl<NodeId, PortName> PortGraphEdge<NodeId, PortName> for ((NodeId, PortName), (NodeId, PortName))
+where
+    NodeId: PortGraphId,
+    PortName: PortGraphId,
+{
+    fn source(&self) -> (&NodeId, &PortName) {
+        (&self.0 .0, &self.0 .1)
+    }
+    fn target(&self) -> (&NodeId, &PortName) {
+        (&self.1 .0, &self.1 .1)
+    }
+}
+
+/// Trait implementing a Port Graph, a directed graph consisting
+/// of vertices that have named input and output ports,
+/// with edges represented by a directed connection from
+/// one vertex's port to another.
+///
+/// See [PortGraph::validate_port_graph] for expected constraints.
+pub trait PortGraph: Debug {
+    /// Graph's node type.
+    type Node: PortGraphNode<Self::NodeId, Self::PortName>;
+
+    /// Graph's edge type.
+    type Edge: PortGraphEdge<Self::NodeId, Self::PortName>;
+
+    /// The type of node's unique identifier.
+    type NodeId: PortGraphId;
+
+    /// The type of port names.
+    type PortName: PortGraphId;
+
+    /// Return all nodes in the graph.
+    fn nodes(&self) -> impl Iterator<Item = &Self::Node>;
+
+    /// Return all edges in the graph.
+    fn edges(&self) -> impl Iterator<Item = &Self::Edge>;
+
+    /// Returns a node with `id` in the graph if found.
+    fn get_node(&self, id: &Self::NodeId) -> Option<&Self::Node>;
+
+    /// Validates the structure of the graph, ensuring
+    /// the following properties:
+    ///
+    /// * Node ids are unique across all nodes
+    /// * Port ids are unique across ports of the
+    ///   same type in a node (two output ports can't have the same id,
+    ///   but an input and output port can).
+    /// * No unconnected nodes.
+    /// * Edges must contain node and port references
+    ///   that exist in the collection of nodes.
+    /// * No input ports with multiple edges (no fan-in).
+    /// * Edges must be unique across the graph. Implicitly
+    ///   enforced by not allowing fan-in.
+    fn validate_port_graph(&self) -> Result<()> {
+        validate_port_graph(self)
+    }
+
+    /// Get all connections for this node's ports. If
+    /// a `port_type` provided, only connections containing
+    /// that type of node port are returned.
+    fn get_connections(&self, node: &Self::Node, port_type: Option<PortType>) -> Vec<&Self::Edge> {
```

Comment:
I wonder if there is a way to imagine this as a `filter_map(...)` on an iterator over edges...

Reply:
:eyes: looking into it, +1 for less allocations